pinot-controller/src/main/java/com/linkedin/pinot/controller/api/storage/PinotSegmentUploadUtils.java

```diff
@@ -0,0 +1,167 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.controller.api.storage;
+
+import com.linkedin.pinot.common.config.TableNameBuilder;
+import com.linkedin.pinot.common.metadata.segment.OfflineSegmentZKMetadata;
+import com.linkedin.pinot.common.metadata.segment.SegmentZKMetadataCustomMapModifier;
+import com.linkedin.pinot.common.segment.SegmentMetadata;
+import com.linkedin.pinot.common.utils.CommonConstants;
+import com.linkedin.pinot.common.utils.FileUploadDownloadClient;
+import com.linkedin.pinot.common.utils.SegmentName;
+import com.linkedin.pinot.controller.ControllerConf;
+import com.linkedin.pinot.controller.api.resources.FileUploadPathProvider;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import com.linkedin.pinot.filesystem.LocalPinotFS;
+import com.linkedin.pinot.filesystem.PinotFS;
+import com.linkedin.pinot.filesystem.PinotFSFactory;
+import java.io.File;
+import java.net.URI;
+import java.util.UUID;
+import javax.inject.Inject;
+import javax.ws.rs.core.HttpHeaders;
+import org.apache.helix.ZNRecord;
+import org.json.JSONException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Contains upload-specific methods
+ */
+public class PinotSegmentUploadUtils {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PinotSegmentUploadUtils.class);
+
+  @Inject
+  PinotHelixResourceManager _pinotHelixResourceManager;
+
+  @Inject
+  ControllerConf _controllerConf;
+
+
+  public URI constructFinalLocation(SegmentMetadata segmentMetadata, SegmentVersion version) throws Exception {
+    FileUploadPathProvider provider = new FileUploadPathProvider(_controllerConf);
+    File destFile = new File(new File(new File(provider.getBaseDataDir(), segmentMetadata.getTableName()), segmentMetadata.getName()), version.toString());
+    return destFile.toURI();
+  }
+  /**
+   * Updates zk metadata for uploaded segments
+   * @param segmentMetadata
+   * @param segmentUploaderConfig
+   */
+  public void pushMetadata(SegmentMetadata segmentMetadata, SegmentUploaderConfig segmentUploaderConfig)
+      throws JSONException {
+    FileUploadPathProvider provider;
+    try {
+      provider = new FileUploadPathProvider(_controllerConf);
+    } catch (Exception e) {
+      LOGGER.error("Controller configuration is misconfigured");
```

Comment:
Always include e.getMessage(); better yet, pass e to the logger to log relevant message, stack trace etc. The error message could also be updated to "Could not create file-upload-provider" instead of concluding that it is an isue with the configuration.